packages/wallets/src/evm/wallet.ts

```diff
@@ -26,48 +68,318 @@ export interface ViemWallet {
 }
 
 export class EVMSmartWallet implements ViemWallet {
+    public readonly publicClient: PublicClient<HttpTransport>;
+
     constructor(
-        public readonly publicClient: PublicClient<HttpTransport>,
         public readonly chain: EVMSmartWalletChain,
         private readonly apiClient: ApiClient,
-        private readonly address: Address
-    ) {}
+        private readonly address: Address,
+        private readonly adminSigner: EVMSigner
+    ) {
+        this.publicClient = createPublicClient({
+            chain: toViemChain(chain),
+            transport: http(),
+        });
+    }
 
-    public async balances() {}
-    public async transactions() {}
-    public async nfts() {}
+    public async balances(tokens: Address[]) {
+        return await this.apiClient.getBalance(this.walletLocator, {
+            chains: [this.chain],
+            tokens,
+        });
+    }
+    public async transactions() {
+        const transactions = await this.apiClient.getTransactions(this.walletLocator);
+        return transactions.transactions.filter((transaction) => transaction.walletType === "evm-smart-wallet");
+    }
+    public async nfts(perPage: number, page: number) {
+        return await this.apiClient.getNfts(this.walletLocator, perPage, page);
+    }
 
     public getAddress() {
         return this.address;
     }
 
-    // biome-ignore lint/suspicious/useAwait: <explanation>
     public async getNonce(parameters?: { key?: bigint | undefined } | undefined): Promise<bigint> {
-        throw new Error("Not implemented");
+        const nonce = await this.publicClient.readContract({
+            abi: entryPointAbi,
+            address: ENTRY_POINT_ADDRESS,
+            functionName: "getNonce",
+            args: [this.address, parameters?.key ?? BigInt(0)],
+        });
+        return nonce;
     }
 
-    // biome-ignore lint/suspicious/useAwait: <explanation>
     public async signMessage(parameters: {
         message: SignableMessage;
     }): Promise<Hex> {
-        throw new Error("Not implemented");
+        if (typeof parameters.message !== "string") {
+            throw new Error("Message must be a string");
+        }
+
+        // Create signature
+        const signatureCreationResponse = await this.apiClient.createSignature(this.walletLocator, {
+            type: "evm-message",
+            params: {
+                message: parameters.message,
+                signer: this.signerLocator,
+                chain: this.chain,
+            },
+        });
+        const signatureId = signatureCreationResponse.id;
+
+        // Approve signature
+        const pendingApprovals = signatureCreationResponse.approvals?.pending || [];
+        if (pendingApprovals.length !== 1) {
+            throw new Error(`Expected 1 pending approval, got ${pendingApprovals.length}`);
+        }
+        const pendingApproval = pendingApprovals[0];
```

Comment:
we should find pending approvals by locator, making sure the signer in the pendingApproval is the same as `this.signerLocator`. Same in transactions and typed data

Reply:
good call https://github.com/Crossmint/crossmint-sdk/pull/970/commits/be3ae7f56923d1c2f6c399dab611f1ce858ce321